feat: 멀티 모듈로 변경

sandropark opened the pull request:
## 작업사항
- 수집, 정제, 색인, 서빙 모듈 생성

## 참고사항


### Review Rule

- P1: 꼭 반영해주세요 (Request changes)
- P2: 적극적으로 고려해주세요 (Request changes)
- P3: 웬만하면 반영해 주세요 (Comment)
- P4: 반영해도 좋고 넘어가도 좋습니다 (Approve)
- P5: 그냥 사소한 의견입니다 (Approve)

sandropark committed:
feat: 멀티모듈 프로젝트 구조 설정

- 모듈 구성: smithy(정제), codex(색인), portal(서빙)
- settings.gradle.kts에서 멀티모듈 선언 추가
- 각 모듈 별 application.yml 및 SpringBootApplication 클래스 추가
- build.gradle.kts 구조 변경: 루트 및 서브프로젝트별 설정 분리
- Gradle 속성 파일 추가 및 Gradle 설정 최적화

sandropark committed:
feat: Harvest 모듈 추가

- 수집용 Harvest 모듈 생성

sandropark committed:
chore: 배치 의존성 추가

sandropark committed:
chore: docker compose로 mysql 생성

sandropark committed:
feat: 스프링 배치로 위키 파일을 읽는 ItemReader 구현

sandropark committed:
docs: 시스템 구상 및 todo list 작성

sandropark committed:
chore: update .gitignore

sandropark committed:
docs: update 시스템 구성도 및 TODO list

sandropark committed:
chore: Harvest 모듈 제거하고 python 프로젝트로 변경 및 설계 문서 추가

sandropark committed:
refactor: Smithy 모듈을 python 프로젝트로 변경

sandropark committed:
chore: docker compose에 MinIO 서비스 추가 및 TODO 리스트 수정

sandropark committed:
chore: docker compose에 Kafka 및 Kafka UI 서비스 추가

sandropark committed:
docs: CLAUDE.md에 프로젝트 구조 및 의존성 정보 업데이트

- 프로젝트를 멀티 모듈로 재구성한 내용 반영
- 각 모듈별 설명 및 의존성 항목 추가
- 테스트 전략 및 코루틴 지원 내용 명시

sandropark committed:
chore: .gitignore에 .env 추가

sandropark committed:
chore: docker-compose.yml 경로 수정 및 healthcheck 배열 스타일 통일

sandropark committed:
chore: 환경 변수 및 의존성 설정 추가

- `.env.example` 파일 추가
- MinIO 및 업로드 로그 관련 환경 변수 설정
- `pyproject.toml`에 `python-dotenv`, `pydantic`, `pydantic-settings` 패키지 추가
- Pydantic 기반 설정 관리 클래스 추가
- TODO 리스트 업데이트

sandropark committed:
feat: 파일을 읽어서 minio에 업로드 로직 작성

sandropark committed:
chore: .gitignore에 업로드 로그 파일 추가

sandropark committed:
docs: TODO 리스트 대용량 파일 업로드 항목 추가

sandropark committed:
docs: TODO 리스트 항목 수정 및 구체화

- 대용량 파일 처리, 뉴스 크롤링, X API 작업 세부 내용 추가
- 중복 항목 제거 및 표현 정리

Seonguk-j commented on docker-compose.yml:
```diff
@@ -0,0 +1,110 @@
+    ports:
+      - "3306:3306"
+    environment:
+      MYSQL_ROOT_PASSWORD: root1234
+      MYSQL_DATABASE: realtime_db
+      MYSQL_USER: realtime
+      MYSQL_PASSWORD: realtime1234
```
연습용 프로젝트라 비밀번호를 따로 빼는건 굳이의 느낌일까요...?
p5

sandropark committed:
feat: 웹 서버 구조로 변경

- FastAPI를 사용

DevRunner21 commented on harvest/src/harvest/upload_wiki_data.py:
```diff
@@ -0,0 +1,224 @@
+logger = logging.getLogger(__name__)
+
+
+class WikiDataUploader:
```
파일 핸들링 하는 부분을 파이썬으로 작성하셨네요!
전 아직 다른 언어 사용해볼 엄두가 안나는데... ㄷㄷ 언어 사용에 유연하시군요! 👍👍

혹시 기존에 사용하시던 코틀린이 아니라 파이썬으로 작성하신 이유가 있을까요??

MinIO에 업로드 한다고 적혀있는데 이게 어떤 건지도 궁급합니다!!

sandropark replied:
파일 핸들링은 솔직히 어떤 언어로 하나 크게 상관없을 것 같은데, 크롤링 부분은 확실히 파이썬이 유리할 것 같아서 선택했습니다. 개발을 파이썬으로 시작해서 내적 친밀감이 있기 때문에 사용한 것도 있습니다 ㅎㅎ

MiniO는 AI한테 물어보니까 AWS S3랑 인터페이스가 동일한 파일 저장소라고 하더라구요. 그래서 추후에 S3로 마이그레이션할 때도 코드 단에서는 변경사항이 최소화될 수 있다 그래서 사용했습니다. raw 데이터는 DB보다는 그냥 파일 저장소에 던져두는 게 편할 것 같아서 선택했습니다. 위키 데이터(.bz2), 뉴스 html, api 응답 json 모두 파일로 만들어서 넣어두려합니다.

DevRunner21 commented on harvest/README.md:
```diff
@@ -0,0 +1,188 @@
+
+### 1. 멀티파트 업로드
+
+- 25MB 이상의 파일은 자동으로 멀티파트 업로드 사용
+- 업로드 중단 시 재시작 가능
+- 병렬 청크 업로드로 성능 향상
```
파일을 청크 단위로 다운로드 하시는 건가요??
그럼 파일 읽을 떄는 어떤식으로 처리하실 예정인지 궁금합니다!

sandropark replied:
이 부분은 AI가 자동 생성해서 어떤 상황인지 모르겠네요 ㅎㅎ;; 정리해야겠네요.

DevRunner21 commented on smithy/smithy.md:
```diff
@@ -0,0 +1,21 @@
+        - newspaper3K를 사용한다.
+      - 해시함수를 태워서 해시값을 redis에 저장. 새로운 페이지 수집 시 같은 방법으로 해시를 비교.
+      - 이미 존재하는 경우 무시 혹은 원본도 삭제
+- 위키
+  - 중복 제거
+    - 문서를 파싱해서 id를 확인
+    - id가 없는 경우: 신규 페이지기 때문에 생성
+    - id가 있는 경우: revisionId를 비교
+        - revisionId가 없는 경우: 새로운 수정본이다. 덮어쓰기
+        - revisionId가 있는 경우: 이미 있는 데이터. 무시하기
```
저 이 부분을 어떻게 처리할까 고민하고 있었는데 ㅋㅋ

저같은 경우 revisionId를 데이터 수집하면서 발라내다보니 중복체크를 수집모듈에서 할 예정이었는데,
산들님의 경우 문서 중복체크에 대한 역할을 정제모듈에 두셨네요!

혹시 이유가 있을까요??
( 뭔가 수집모듈에서는 수집만 딱! 하는게 깔끔해보여서 저도 책임을 옮겨볼까 고민됩니다! )

sandropark replied:
저도 처음에는 수집 시스템에서 중복을 제어할까 싶었는데 그러려면 위키 데이터 특성 상 파일을 파싱하는 작업까지해야 돼서 그냥 정제 모듈로 넘겼습니다. 어차피 정제할 때 파싱해야 하니까 그때 처리하려구요 ㅎ

추가로 수집에서 중복을 제어하려면 결국 위키 데이터는 <page> 단위로 저장이 될 텐데 파일하나에 10만+ 개의 페이지가 있으니까   너무 과하다고 생각해서 빼기도 했습니다.

DevRunner21 reviewed (commented):
안녕하세요! 리뷰 드렸습니다!
전반적으로 리드미를 너무 잘 작성해주셔서 보기가 수월했습니다!
( 저도 이렇게 적어야하는데... 빨리 정리해보겠습니다.. ㅜㅜ )

파일처리 부분을 파이썬으로 작성하신 부분이 인상적이었어요!
파이썬을 선택하신 이유가 어떤건지 궁금합니다!

메일보는게 늦어서 리뷰 드리는게 늦었습니다 ㅜㅜ
다음엔 바로 리뷰 올릴께요!!

sandropark commented:
> 안녕하세요! 리뷰 드렸습니다! 전반적으로 리드미를 너무 잘 작성해주셔서 보기가 수월했습니다! ( 저도 이렇게 적어야하는데... 빨리 정리해보겠습니다.. ㅜㅜ )
> 
> 파일처리 부분을 파이썬으로 작성하신 부분이 인상적이었어요! 파이썬을 선택하신 이유가 어떤건지 궁금합니다!
> 
> 메일보는게 늦어서 리뷰 드리는게 늦었습니다 ㅜㅜ 다음엔 바로 리뷰 올릴께요!!

감사합니다! 아직 작업 초기 단계라 아마 계속 바뀔 것 같습니다 ㅎㅎ;; 그냥 pr 올려뒀는데 리뷰 주셔서 감사합니다.

sandropark committed:
chore: pyyaml 패키지 의존성 추가

sandropark committed:
feat: minio, kafka client 수정

sandropark committed:
feat: 업로드 API 구현

Seonguk-j reviewed (approved):
코틀린과 파이썬이라 제가 잘 이해하지 못해 코멘트를 남기기가 어렵네요... 죄송합니다..

sandropark committed:
feat: X API 클라이언트 및 서비스 구현

- X API 클라이언트 클래스 작성
- 사용자 트윗 검색 및 타임라인 조회 기능 추가
- 관련 스키마, 예외 클래스, 라우트 정의
- FastAPI에 X API 관련 엔드포인트 추가
- `.env.example`에 X API 환경 변수 설정 추가
- Tweepy 패키지 의존성 추가

sandropark committed:
feat: Google News 클라이언트 및 테스트 구현

- Google News RSS 클라이언트 작성
- 뉴스 카테고리 Enum 클래스 추가
- Pydantic 기반 뉴스 스키마 정의
- 클라이언트 및 뉴스 스키마 관련 유닛 테스트 작성
- 통합 테스트 코드 작성
- 기존 불필요한 imports 제거 및 지연 import로 성능 최적화

sandropark committed:
feat: Google News 클라이언트 및 테스트 구현

- Google News RSS 클라이언트 작성
- 뉴스 카테고리 Enum 클래스 추가
- Pydantic 기반 뉴스 스키마 정의
- 클라이언트 및 뉴스 스키마 관련 유닛 테스트 작성
- 통합 테스트 코드 작성
- 기존 불필요한 imports 제거 및 지연 import로 성능 최적화

sandropark committed:
Merge remote-tracking branch 'origin/feature/2025-09-02/multi-module' into feature/2025-09-02/multi-module

# Conflicts:
#	harvest/tests/test_news.py

sandropark committed:
feat: Elasticsearch 연동

sandropark committed:
refactor: 게시글 관련 기능 제거 및 문서 관리 기능 추가

- 게시글 도메인, 컨트롤러, 서비스 및 테스트 코드 제거
- 문서 관리 서비스 및 관련 로직 추가
- 불필요한 파일 및 코드 정리

sandropark committed:
feat: 카프카 프로듀서 및 문서 메시지 발행

sandropark committed:
feat: 카프카 컨슈머 추가
